bip-0324.mediawiki

```diff
@@ -0,0 +1,467 @@
+ |               <-- SERVER_HDATA ----                                                  |
+ |                                                                                      |
+ | ECDH_KEY     := SECP256k1_ECDH(x,Y)                                                  |
+ ----------------------------------------------------------------------------------------
+</pre>
+
+To request encrypted communication (only possible if yet no other messages have been sent or received), the initiating peer generates an EC secp256k1 ephemeral key and sends the corresponding 32-byte public key to the responding peer and waits for the remote 32-byte public key from the counterparty.
+
+ODD secp256k1 public keys MUST be used (public keys starting with 0x02). If the public key from the generated ephemeral key is an EVEN public key (starting with 0x03), its public key SHOULD be negated and then recalculated.
+Only using ODD public keys makes it more complex to identify the handshake based on analyzing the traffic.
+
+The handshake request and response message are raw 32byte payloads containing no header, length or checksum (the pure 32byte payload) and MUST be sent before anything else.
+
+Public keys starting with the 4-byte network magic are forbidden and MUST lead to local regeneration of an ephemeral-key.
+
+Pseudocode for the ephemeral-key generation
+<pre>
+do {
+    ecdh_key.MakeNewKey();
+    if (ecdh_key.GetPubKey()[0] == 3) {
+        ecdh_key.Negate();
+    }
+} while (m_ecdh_key.GetPubKey()[0..3] == NETWORK_MAGIC);
+</pre>
+
+Once a peer has received the public key from its counterparty, the shared secret MUST be calculated by using secp256k1 ECDH.
+
+Private keys will never be transmitted. The shared secret can only be calculated if an attacker knows at least one private key and the counterparty's public key. This key-exchange is based on the discrete log problem and thus not sufficiently strong against known forms of possible quantum computer algorithms. Adding an additional quantum resistant key exchange like NewHope is possible but out of scope for this proposal.
+
+After a successful handshake, messages MUST use the "v2 messages structure". Non-encrypted v1 messages from the initiating peer MUST lead to an immediate connection termination.
+
+After a successful handshake, both peers MUST wipe the ephemeral-session-key from memory and/or persistence storage.
+
+A peer not supporting this proposal will not perform the described handshake and thus send a v1 version message.
+Peers supporting this BIP MAY optionally allow unencrypted v1 communication by detecting a v1 version message by the initial 11-byte sequence of <code>4byte net magic || "version"</code>.
+
+=== Symmetric Encryption Cipher Keys ===
+
+Once the ECDH secret (<code>ECDH_KEY</code>) is calculated on each side, the symmetric encryption cipher keys MUST be derived with HKDF <ref>[https://tools.ietf.org/html/rfc5869 HKDF (RFC 5869)]</ref> after the following specification:
```

Comment:
* "Once the secret is calculated on each side" -> "Once the secret has been calculated on both sides"
* "with HKDF" -> "with the HKDF"
* "after the following specification" -> "using the following specification"

These are just my suggestions, I think it is easier to read but a native speaker like @jnewbery could confirm.